File: mobile/android/app/src/main/kotlin/com/hearth/ui/QuickMeasurement.kt
package com.hearth.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedCard
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import com.hearth.data.ConsentKind
import com.hearth.data.HealthEntry
import com.hearth.data.HealthKind
import com.hearth.data.NewHealthEntry
import com.hearth.data.Person
import com.hearth.data.Repo
import com.hearth.health.findPreset
import com.hearth.health.formatValue
import com.hearth.health.localDate
import com.hearth.health.measurementOrder
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.time.LocalTime

private const val CUSTOM = ""

private fun number(s: String) = s.trim().replace(',', '.').toDoubleOrNull()?.takeIf { it.isFinite() }

/**
 * One line for the numbers people record over and over (QuickMeasurement.tsx): pick the parameter,
 * type the number, save. Date and time are now, the unit comes from the preset, and the parameters
 * this person already records come first. Anything more belongs in the full form.
 */
@Composable
fun QuickMeasurement(repo: Repo, persons: List<Person>, initialPerson: String, entries: List<HealthEntry>, onSaved: () -> Unit) {
    val t = LocalStrings.current
    val scope = rememberCoroutineScope()
    var personId by remember(initialPerson) { mutableStateOf(initialPerson) }
    val order = remember(entries, personId) { measurementOrder(entries.filter { personId.isEmpty() || it.personId == personId }) }
    var presetId by remember { mutableStateOf(order.firstOrNull()?.id ?: CUSTOM) }
    var title by remember { mutableStateOf("") }
    var unit by remember { mutableStateOf("") }
    var value by remember { mutableStateOf("") }
    var value2 by remember { mutableStateOf("") }
    var asking by remember { mutableStateOf(false) }
    var saved by remember { mutableStateOf<String?>(null) }

    val preset = findPreset(presetId)
    val pair = preset?.pair
    val finalTitle = preset?.title ?: title.trim()
    val finalUnit = preset?.unit ?: unit.trim()
    val valid = personId.isNotEmpty() && finalTitle.isNotEmpty() && number(value) != null && (pair == null || number(value2) != null)

    fun save() = scope.launch {
        if (!valid) return@launch
        if (!withContext(Dispatchers.IO) { repo.hasConsent(ConsentKind.IMPORT_DOCUMENT, personId) }) {
            asking = true
            return@launch
        }
        val v = number(value)
        val v2 = if (pair != null) number(value2) else null
        withContext(Dispatchers.IO) {
            repo.addHealthEntry(
                NewHealthEntry(
                    personId, localDate(), LocalTime.now().let { "%02d:%02d".format(it.hour, it.minute) }, HealthKind.MEASUREMENT,
                    finalTitle, "", preset?.bodyPart ?: "", null, preset?.tags ?: emptyList(), v, v2, finalUnit,
                ),
            )
        }
        saved = "$finalTitle ${formatValue(v, v2, finalUnit)}"
        value = ""
        value2 = ""
        onSaved()
    }

    OutlinedCard(Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 8.dp)) {
        Column(Modifier.padding(12.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
            Text(t("quick.label"), style = MaterialTheme.typography.titleSmall)
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                if (persons.size > 1) {
                    Picker(t("healthPage.colPerson"), personId, listOf("" to t("healthForm.pickPerson")) + persons.map { it.id to it.displayName }, modifier = Modifier.weight(1f)) { personId = it }
                }
                Picker(t("quick.parameter"), presetId, order.map { it.id to t("preset.${it.id}") } + (CUSTOM to t("quick.custom")), modifier = Modifier.weight(1f)) {
                    presetId = it
                    value = ""
                    value2 = ""
                }
            }
            if (preset == null) {
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    OutlinedTextField(title, { title = it }, placeholder = { Text(t("quick.customPlaceholder")) }, singleLine = true, modifier = Modifier.weight(1f))
                    OutlinedTextField(unit, { unit = it }, placeholder = { Text(t("healthLog.unitPlaceholder")) }, singleLine = true, modifier = Modifier.weight(0.6f))
                }
            }
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp), verticalAlignment = Alignment.CenterVertically) {
                val keyboard = KeyboardOptions(keyboardType = KeyboardType.Decimal, imeAction = ImeAction.Done)
                OutlinedTextField(
                    value, { value = it },
                    placeholder = { Text(if (pair != null) t("preset.pair.${pair.first}") else t("healthLog.value")) },
                    singleLine = true, keyboardOptions = keyboard, keyboardActions = KeyboardActions(onDone = { save() }), modifier = Modifier.weight(1f),
                )
                if (pair != null) {
                    OutlinedTextField(
                        value2, { value2 = it },
                        placeholder = { Text(t("preset.pair.${pair.second}")) },
                        singleLine = true, keyboardOptions = keyboard, keyboardActions = KeyboardActions(onDone = { save() }), modifier = Modifier.weight(1f),
                    )
                }
                if (preset != null) Text(preset.unit, style = MaterialTheme.typography.bodyMedium, color = MaterialTheme.colorScheme.onSurfaceVariant)
                Button(enabled = valid, onClick = { save() }) { Text(t("quick.save")) }
            }
            saved?.let { Text(t("quick.saved", "what" to it), style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.primary) }
        }
    }

    if (asking) {
        val ticked = rememberTicks(ConsentKind.IMPORT_DOCUMENT)
        AlertDialog(
            onDismissRequest = { asking = false },
            title = { Text(t(ConsentKind.IMPORT_DOCUMENT.titleKey)) },
            text = { ConsentChecks(ConsentKind.IMPORT_DOCUMENT, ticked) },
            confirmButton = {
                TextButton(enabled = ticked.all { it }, onClick = {
                    asking = false
                    scope.launch {
                        withContext(Dispatchers.IO) { repo.grantConsent(ConsentKind.IMPORT_DOCUMENT, personId) }
                        save()
                    }
                }) { Text(t("consentForm.confirm")) }
            },
            dismissButton = { TextButton(onClick = { asking = false }) { Text(t("common.cancel")) } },
        )
    }
}

File: mobile/ios/Hearth/Native/UI/QuickMeasurement.swift
import SwiftUI

/// One line for the numbers people record over and over (QuickMeasurement.tsx): pick the parameter,
/// type the number, save. Date and time are now, the unit comes from the preset, and the parameters
/// this person already records come first. Anything more belongs in the full form.
struct QuickMeasurement: View {
    @ObservedObject var store: HealthStore
    /// The person in scope, or "" to pick one here.
    let initialPerson: String

    /// The tag of "something else": a title and unit typed here.
    private static let custom = ""

    @State private var personId = ""
    @State private var presetId: String?
    @State private var title = ""
    @State private var unit = ""
    @State private var value = ""
    @State private var value2 = ""
    @State private var asking = false
    @State private var ticked = ConsentChecks.unticked(.importDocument)
    @State private var saved: String?
    @State private var failure: String?

    private var order: [HealthPreset] {
        HealthPresets.measurementOrder(store.entries.filter { personId.isEmpty || $0.personId == personId })
    }

    private var selectedId: String { presetId ?? order.first?.id ?? QuickMeasurement.custom }
    private var preset: HealthPreset? { HealthPresets.find(selectedId) }
    private var pair: [String]? { preset?.pair }
    private var finalTitle: String { preset?.title ?? title.trimmingCharacters(in: .whitespacesAndNewlines) }
    private var finalUnit: String { preset?.unit ?? unit.trimmingCharacters(in: .whitespacesAndNewlines) }

    private var valid: Bool {
        !personId.isEmpty && !finalTitle.isEmpty && QuickMeasurement.number(value) != nil
            && (pair == nil || QuickMeasurement.number(value2) != nil)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(t("quick.label")).font(.subheadline.weight(.semibold))
            HStack {
                if store.persons.count > 1 {
                    Picker(t("healthPage.colPerson"), selection: $personId) {
                        Text(t("healthForm.pickPerson")).tag("")
                        ForEach(store.persons) { p in Text(p.displayName).tag(p.id) }
                    }
                }
                Picker(t("quick.parameter"), selection: Binding(
                    get: { selectedId },
                    set: { id in
                        presetId = id
                        value = ""
                        value2 = ""
                    }
                )) {
                    ForEach(order) { p in Text(t("preset.\(p.id)")).tag(p.id) }
                    Text(t("quick.custom")).tag(QuickMeasurement.custom)
                }
            }
            .pickerStyle(.menu)
            if preset == nil {
                HStack {
                    TextField(t("quick.customPlaceholder"), text: $title)
                    TextField(t("healthLog.unitPlaceholder"), text: $unit).frame(maxWidth: 100)
                }
                .textFieldStyle(.roundedBorder)
            }
            HStack {
                TextField(pair.map { t("preset.pair.\($0[0])") } ?? t("healthLog.value"), text: $value)
                    .keyboardType(.decimalPad)
                if let pair, pair.count > 1 {
                    TextField(t("preset.pair.\(pair[1])"), text: $value2)
                        .keyboardType(.decimalPad)
                }
                if let preset, let unit = preset.unit {
                    Text(unit).foregroundStyle(.secondary)
                }
                Button(t("quick.save")) { save() }
                    .buttonStyle(.borderedProminent)
                    .disabled(!valid)
            }
            .textFieldStyle(.roundedBorder)
            if let saved {
                Text(t("quick.saved", ["what": saved])).font(.caption).foregroundStyle(.tint)
            }
            if let failure {
                Text(failure).font(.caption).foregroundStyle(.red)
            }
        }
        .padding(12)
        .background(
            Color(uiColor: .secondarySystemGroupedBackground),
            in: RoundedRectangle(cornerRadius: 12, style: .continuous)
        )
        .onAppear { if personId.isEmpty { personId = initialPerson } }
        .onChange(of: initialPerson) { _, id in personId = id }
        .sheet(isPresented: $asking) { consentSheet }
    }

    /// The person's first entry asks for the document consent, as the full form does.
    private var consentSheet: some View {
        NavigationStack {
            Form {
                ConsentChecks(kind: .importDocument, ticked: $ticked, header: t(ConsentKind.importDocument.titleKey))
                Section {
                    Button(t("consentForm.confirm")) {
                        do {
                            try store.grantDocumentConsent(personId)
                            asking = false
                            save()
                        } catch {
                            failure = t("native.error", ["error": error.localizedDescription])
                        }
                    }
                    .disabled(!ConsentChecks.allTicked(ticked, .importDocument))
                }
            }
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button(t("common.cancel")) { asking = false }
                }
            }
        }
        .presentationDetents([.medium, .large])
    }

    /// A number as typed: either decimal separator, since a decimal pad offers the locale's own.
    private static func number(_ text: String) -> Double? {
        let cleaned = text.trimmingCharacters(in: .whitespaces).replacingOccurrences(of: ",", with: ".")
        guard !cleaned.isEmpty, let n = Double(cleaned), n.isFinite else { return nil }
        return n
    }

    private func save() {
        guard valid else { return }
        do {
            guard try store.hasDocumentConsent(personId) else {
                ticked = ConsentChecks.unticked(.importDocument)
                asking = true
                return
            }
            let v = QuickMeasurement.number(value)
            let v2 = pair != nil ? QuickMeasurement.number(value2) : nil
            let clock = Calendar.current.dateComponents([.hour, .minute], from: Date())
            let entry = try store.add(HealthEntry(
                id: "", personId: personId, date: HealthLog.localDate(),
                time: String(format: "%02ld:%02ld", clock.hour ?? 0, clock.minute ?? 0),
                kind: .measurement, title: finalTitle, body: "", source: "", bodyPart: preset?.bodyPart ?? "",
                severity: nil, tags: preset?.tags ?? [], value: v, value2: v2, unit: finalUnit, createdAt: ""
            ))
            saved = "\(finalTitle) \(HealthLog.formatValue(entry))"
            value = ""
            value2 = ""
            failure = nil
        } catch {
            failure = t("native.error", ["error": error.localizedDescription])
        }
    }
}
